core/network/impl/stream_engine.hpp

```diff
@@ -261,13 +264,59 @@ namespace kagome::network {
       });
     }
 
-    template <typename T>
-    void updateStream(const PeerInfo &peer,
+    template <typename F, typename H>
+    void forNewStream(const PeerInfo &peer,
                       const Protocol &protocol,
-                      std::shared_ptr<T> msg) {
+                      boost::optional<std::shared_ptr<H>> handshake,
+                      F &&f) {
+      using CallbackResultType =
+          outcome::result<std::shared_ptr<libp2p::connection::Stream>>;
       host_.newStream(
           peer,
           protocol,
+          [peer, handshake{std::move(handshake)}, f{std::forward<F>(f)}](
+              auto &&stream_res) mutable {
+            if (!stream_res || !handshake) {
```

Comment:
Could you please comment this case? It's not rightaway clear for me what's going on here, so I think a comment would be of help in the future.

Reply:
This code will be refactored as well. It meaning is very simple: we execute the logic below only in the case we have no errors and valid handshake data to process.